internal/deployment_client.go

```diff
@@ -0,0 +1,225 @@
+		// Name - Task queue name.
+		Name string
+
+		// Type - The type of this task queue.
+		Type TaskQueueType
+
+		// FirstPollerTime - Time when the server saw the first poller for this task queue
+		// in this deployment.
+		FirstPollerTime time.Time
+	}
+
+	// DeploymentInfo holds information associated with
+	// workers in this deployment.
+	// Workers can poll multiple task queues in a single deployment,
+	// which are listed in this message.
+	// NOTE: Experimental
+	DeploymentInfo struct {
+		// Deployment - An identifier for this deployment.
+		Deployment Deployment
+
+		// CreateTime - When this deployment was created.
+		CreateTime time.Time
+
+		// Current - Whether this deployment is the current one for its deployment series.
+		Current bool
+
+		// TaskQueuesInfo - List of task queues polled by workers in this deployment.
+		TaskQueuesInfo []DeploymentTaskQueueInfo
+
+		// Metadata - A user-defined set of key-values. Can be updated with `DeploymentClient.SetCurrent`.
+		Metadata map[string]*commonpb.Payload
+	}
+
+	// DeploymentListEntry is a subset of fields from DeploymentInfo
+	// NOTE: Experimental
+	DeploymentListEntry struct {
+		// An identifier for this deployment.
+		Deployment Deployment
+
+		// When this deployment was created.
+		CreateTime time.Time
+
+		// Whether this deployment is the current one for its deployment series.
+		Current bool
+	}
+
+	// DeploymentListIterator is an iterator for deployments.
+	// NOTE: Experimental
+	DeploymentListIterator interface {
+		// HasNext - Return whether this iterator has next value.
+		HasNext() bool
+
+		// Next - Returns the next deployment and error
+		Next() (*DeploymentListEntry, error)
+	}
+
+	// DeploymentListOptions are the parameters for configuring listing deployments.
+	// NOTE: Experimental
+	DeploymentListOptions struct {
+		// PageSize - How many results to fetch from the Server at a time.
+		// Optional: defaulted to 1000
+		PageSize int
+
+		// SeriesName - Filter with the name of the deployment series.
+		// Optional: If present, use an exact series name match.
+		SeriesName string
+	}
+
+	// DeploymentReachabilityInfo extends DeploymentInfo with reachability information.
+	// NOTE: Experimental
+	DeploymentReachabilityInfo struct {
+		DeploymentInfo
+
+		// Reachability - Kind of tasks that may reach a worker
+		// associated with a deployment.
+		Reachability DeploymentReachability
+
+		// LastUpdateTime - When reachability was last computed. Computing reachability
+		// is an expensive operation, and the server caches results.
+		LastUpdateTime time.Time
+	}
+
+	// DeploymentMetadataUpdate modifies user-defined metadata entries that describe
+	// a deployment.
+	// NOTE: Experimental
+	DeploymentMetadataUpdate struct {
+		// UpsertEntries - Metadata entries inserted or modified. When values are not
+		// of type *commonpb.Payload, the default data converter will be used to generate
+		// payloads.
+		UpsertEntries map[string]interface{}
```

Comment:
> the default data converter will be used to generate payloads.

Why? does the Temporal server need to read these?

Reply:
The idea of payloads was to support encryption. Converting to payloads transparently in UpsertEntries was just for convenience, they can always pass payloads if they want...

Reply:
We should use the clients data converter here then

Reply:
We should only use the default if the server needs to read the payload

Reply:
I'm using `converter.GetDefaultDataConverter().ToPayload`, I thought that would give me client data converter, which one should I be using?

Reply:
I see `getDataConverterFromWorkflowContext()` may give me what I need, it defaults to the default data converter if none provided...

Reply:
`WorkflowClient.dataConverter`

Reply:
`GetDefaultDataConverter()` is a program wide data converter, not the clients

Reply:
Server does not need to read these.

Reply:
Done